WifiDirectTest/app/src/main/java/test/microsoft/com/wifidirecttest/WifiDirectService.java

```diff
@@ -0,0 +1,411 @@
+
+    private final IBinder mBinder = new LocalBinder();
+    public class LocalBinder extends Binder {
+        WifiDirectService getService() {
+            return WifiDirectService.this;
+        }
+    }
+    @Override
+    public IBinder onBind(Intent intent) {
+        return mBinder;
+    }
+
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        return START_STICKY;
+    }
+
+    @Override
+    public void onCreate() {
+        mySpeech = new MyTextSpeech(this);
+
+        broadcaster = LocalBroadcastManager.getInstance(this);
+
+
+        p2p = (WifiP2pManager) getSystemService(WIFI_P2P_SERVICE);
+        if (p2p == null) {
+            mySpeech.speak("This device does not support Wi-Fi Direct");
+            sendResult("This device does not support Wi-Fi Direct");
+        } else {
+            channel = p2p.initialize(this, getMainLooper(), null);
+
+            receiver = new PeerReceiver();
+            filter = new IntentFilter();
+            filter.addAction(WIFI_P2P_STATE_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_PEERS_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_CONNECTION_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+            registerReceiver(receiver, filter);
+
+            openWifiSocket();
+
+            peerListListener = new WifiP2pManager.PeerListListener() {
+
+                public void onPeersAvailable(WifiP2pDeviceList peers) {
+                    sendResult("Discovered peers:");
+
+                    final WifiP2pDeviceList pers = peers;
+                    String allInOne = "";
+                    int numm = 0;
+                    for (WifiP2pDevice peer : pers.getDeviceList()) {
+                        numm++;
+                        allInOne = allInOne + MyP2PHelper.deviceToString(peer) + ", ";
+                        sendResult("\t" + MyP2PHelper.deviceToString(peer));
+                    }
+                    mySpeech.speak(numm + " peers discovered, names: " + allInOne);
+                }
+            };
+
+            serviceListener = new WifiP2pManager.DnsSdServiceResponseListener() {
+
+                public void onDnsSdServiceAvailable(String instanceName, String serviceType, WifiP2pDevice device) {
+                    sendResult("\t\n Service discovered, " + instanceName + " " + serviceType + " : " + MyP2PHelper.deviceToString(device));
+                    mySpeech.speak(" Service discovered, " + instanceName + " " + serviceType + " : " + MyP2PHelper.deviceToString(device));
+                }
+            };
+
+            txtListener = new WifiP2pManager.DnsSdTxtRecordListener() {
+
+                public void onDnsSdTxtRecordAvailable(String domain, Map<String, String> txtMap, WifiP2pDevice device) {
+
+                    // should show this on ui
+
+                    final String hjepper = "" + MyP2PHelper.deviceToString(device) + " : " + domain;
```

Comment:
What is hjepper? Online translators think it's Norwegian but won't translate it. :)